Standalone kdl_parser library

christian-rauch opened the pull request:
As proposed in https://github.com/ros/kdl_parser/issues/12, this PR makes `catkin` and `rosconsole` optional. I.e. plain CMake macros are used if catkin is not available and logging is done via std io streams if `rosconsole` is not available.

Since [`urdf`](https://github.com/ros/urdf) itself depends on ROS for similar reasons, it has been replaced by `urdfdom`. This disables some of the functions like `treeFromParam` and `treeFromXml`, which use methods that are not present in `urdfdom`. In the log term it would be beneficial, if `urdf` also would be disentangled from ROS.

I copied `FindTinyXML.cmake` and `FindTinyXML2.cmake` from https://github.com/ros/cmake_modules into this repo, to not introduce another dependency.

christian-rauch committed:
remove boost dependency

christian-rauch committed:
make catkin and rosconsole optional, replace urdf by urdfdom

christian-rauch committed:
cmake modules for TinyXML and TinyXML2

source: https://github.com/ros/cmake_modules

christian-rauch commented:
@sloretz Can you comment on this PR?

sloretz commented on kdl_parser/CMakeLists.txt:
```diff
@@ -2,43 +2,71 @@ cmake_minimum_required(VERSION 2.8.3)
 )
+
+if(NOT catkin_FOUND)
+  set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
```
`list(APPEND CMAKE_MODULE_PATH ${CMAKE_CURRENT_LIST_DIR}/cmake)` so as not to overwrite a path that's been manually appended to

sloretz commented on kdl_parser/cmake/FindTinyXML.cmake:
```diff
@@ -0,0 +1,74 @@
+##################################################################################################
+# 
+# CMake script for finding TinyXML.
```
I was going to say it looks like this [was copied from urdfdom](https://github.com/ros/urdfdom/blob/master/cmake/FindTinyXML.cmake), in which case this shouldn't be needed since `find_package(urdfdom REQUIRED)` is used in the `CMakeLists.txt`, but it looks like `urdfdom` is not `find_package`ing TinyXML in its `<project>-config.cmake` [like it should](https://github.com/ros/urdfdom/blob/1857a55897295c1fa46669bf1483dc78cf2467a8/urdf_parser/include/urdf_parser/urdf_parser.h#L42)

sloretz replied:
ros/urdfdom#120

sloretz commented on kdl_parser/cmake/FindTinyXML2.cmake:
```diff
@@ -0,0 +1,74 @@
+##################################################################################################
+# 
+# CMake script for finding TinyXML2.
```
Just noting it looks like this was copied from https://github.com/ros/cmake_modules/blob/0.4-devel/cmake/Modules/FindTinyXML2.cmake

christian-rauch replied:
Yes, I copied it from the `cmake_modules` repo to be used as standalone. I mentioned the source in the git commit. Should I leave a note in the CMakeLists.txt file where the cmake modules are from?

sloretz replied:
> Should I leave a note in the CMakeLists.txt file where the cmake modules are from?

Yes please

sloretz commented on kdl_parser/include/kdl_parser/kdl_parser.hpp:
```diff
@@ -77,8 +79,10 @@ bool treeFromString(const std::string & xml, KDL::Tree & tree);
  * \param[out] tree The resulting KDL Tree
  * \return true on success, false on failure
  */
+#ifdef HAS_URDF
```
I would rather keep the signature, but return `False` if the library was not built with urdf. That way a downstream package including `kdl_parser.hpp` doesn't have to define `HAS_URDF` themselves.

@clalancette thoughts?

sloretz commented on kdl_parser/src/kdl_parser.cpp:
```diff
@@ -86,8 +93,11 @@ KDL::Joint toKdl(urdf::JointSharedPtr jnt)
+#ifdef HAS_ROS
+        ROS_WARN("Converting unknown joint type of joint '%s' into a fixed joint", jnt->name.c_str());
+#else
+        std::cerr << "Converting unknown joint type of joint '" << jnt->name << "' into a fixed joint" << std::endl;
```
Recommend defining at the top of the file `ROS_WARN`/`ROS_ERROR`/etc instead to keep the number of `HAS_ROS` checks to a minimum.

christian-rauch replied:
Good idea. I forwarded warnings and errors to stderr.

sloretz commented on kdl_parser/src/kdl_parser.cpp:
```diff
@@ -156,56 +168,97 @@ bool addChildrenToTree(urdf::LinkConstSharedPtr root, KDL::Tree & tree)
   return treeFromXml(&urdf_xml, tree);
+#else
+  const urdf::ModelInterfaceSharedPtr robot_model = urdf::parseURDFFile(file);
+  return kdl_parser::treeFromUrdfModel(*robot_model, tree);
```
If the result is the same, how about this always uses the non-urdf version?

sloretz commented on kdl_parser/src/kdl_parser.cpp:
```diff
@@ -156,56 +168,97 @@ bool addChildrenToTree(urdf::LinkConstSharedPtr root, KDL::Tree & tree)
   return treeFromXml(&urdf_xml, tree);
+#else
+  const urdf::ModelInterfaceSharedPtr robot_model = urdf::parseURDF(xml);
+  return kdl_parser::treeFromUrdfModel(*robot_model, tree);
```
Same comment, if the result is the same how about always using this version?

sloretz commented on kdl_parser/src/kdl_parser.cpp:
```diff
@@ -156,56 +168,97 @@ bool addChildrenToTree(urdf::LinkConstSharedPtr root, KDL::Tree & tree)
+  urdf::Model robot_model;
   if (!robot_model.initString(ss.str())) {
+#else
+  const urdf::ModelInterfaceSharedPtr robot_model = urdf::parseURDF(ss.str());
```
same comment about always using urdfdom version

sloretz reviewed (commented):
Thanks for the PR, and your patience @christian-rauch . Comments are below.

christian-rauch committed:
append to cmake module path

christian-rauch commented:
Thanks for coming back to this PR.

I discovered that the non-catkin version doesn't build any more.
The urdf lib version in ROS melodic (1.13) uses TinyXML2, i.e. `initXml (const tinyxml2::XMLDocument *xml)`, while the Ubuntu repo (version 1.12) still uses TinyXML(1), i.e. `initXml (TiXmlElement *xml)`, like in ROS kinetic.
I might have intended to base this PR on kinetic-devel instead of melodic-devel.

Anyway, should we maybe use `urdf::parseURDF` from (urdf_parser / liburdfdom-dev) instead of the manual XML loading and parsing?

christian-rauch committed:
replace urdf by urdfdom

christian-rauch commented:
I went ahead and replaced most of the `urdf` library with `urdfdom`. The urdf library is basically only required if you want to read a URDF model from a ROS parameter. Maybe this could also replaced by ROS methods to remove the `urdf` library dependency completely.

sloretz commented on kdl_parser/include/kdl_parser/kdl_parser.hpp:
```diff
@@ -61,8 +61,10 @@ bool treeFromFile(const std::string & file, KDL::Tree & tree);
  * \param tree The resulting KDL Tree
  * returns true on success, false on failure
  */
+#if defined(HAS_ROS) && defined(HAS_URDF)
```
Mind removing `HAS_ROS` and `HAS_URDF` from the header file? If I understand correctly all downstream ROS users will need to define `HAS_ROS` and `HAS_URDF` before this file is included if they want to use this method. Instead I think`treeFromParam()` can return `false` if it was built without ROS support.

sloretz commented on kdl_parser/CMakeLists.txt:
```diff
@@ -2,43 +2,74 @@ cmake_minimum_required(VERSION 2.8.3)
+  )
+endif()
 
 add_library(${PROJECT_NAME} src/kdl_parser.cpp)
```
It looks like there's some catkin magic here. Where when I build with `catkin_make` it creates a shared library, but when I build without ROS on ubuntu 18.04 it builds a static library. Maybe add `SHARED` to this call?

note to self for non-ros building
```
CMAKE_PREFIX_PATH=/opt/ros/melodic/share/orocos_kdl/cmake/ cmake ../src/kdl_parser/kdl_parser
```

christian-rauch replied:
Good catch. It's shared now.

christian-rauch committed:
mention source of local cmake module files

christian-rauch committed:
define ROS_WARN and ROS_ERROR as standard error

christian-rauch committed:
remove define guards from header files

christian-rauch committed:
always build shared library

sloretz reviewed (approved):
LGTM! @clalancette Mind taking a look at this one too?

I ran `check_kdl_parser` on `test/pr2_desc.xml` building with ROS and without. The only difference I see is how `ROS_WARN`/`ROS_ERROR` are output, which is fine.

When building with ROS `ROS_WARN` is output to stdout with formatting that includes a timestamp and console colors. When building without ROS the warning is sent to stderr and is just the text of the warning itself.

christian-rauch commented:
@sloretz @clalancette Any news/feedback on this?

sloretz commented:
ping @clalancette , ok to merge this one?

clalancette commented on kdl_parser/src/kdl_parser.cpp:
```diff
@@ -129,7 +141,9 @@ KDL::RigidBodyInertia toKdl(urdf::InertialSharedPtr i)
 {
   std::vector<urdf::LinkSharedPtr> children = root->child_links;
+#ifdef HAS_ROS
   ROS_DEBUG("Link %s had %zu children", root->name.c_str(), children.size());
```
Given that we have overrides at the top for `ROS_ERROR` and `ROS_WARN`, it seems like we should have one for `ROS_DEBUG` as well.

christian-rauch replied:
`ROS_DEBUG` is now forwarded to `stdout`.
I think I initially did not define it to not clutter the output, since it should not be `stderr` and since `stdout` cannot be inhibited without removing all other `stdout` from a process.

clalancette reviewed (approved):
One minor change, but I'll still approve.

christian-rauch committed:
define ROS_DEBUG as standard output

sloretz commented:
Thanks for the PR!
